AI/VCAI/Fuzzy.cpp

```diff
@@ -373,142 +373,93 @@ float FuzzyHelper::evaluate(Goals::RecruitHero & g)
 {
 	return 1;
 }
-HeroMovementGoalEngine::~HeroMovementGoalEngine()
-{
-	delete strengthRatio;
-	delete heroStrength;
-	delete turnDistance;
-	delete missionImportance;
-	delete estimatedReward;
-}
-
-void FuzzyHelper::initVisitTile()
+HeroMovementGoalEngineBase::HeroMovementGoalEngineBase()
 {
 	try
 	{
-		vt.strengthRatio = new fl::InputVariable("strengthRatio"); //hero must be strong enough to defeat guards
-		vt.heroStrength = new fl::InputVariable("heroStrength"); //we want to use weakest possible hero
-		vt.turnDistance = new fl::InputVariable("turnDistance"); //we want to use hero who is near
-		vt.missionImportance = new fl::InputVariable("lockedMissionImportance"); //we may want to preempt hero with low-priority mission
-		vt.estimatedReward = new fl::InputVariable("estimatedReward"); //indicate AI that content of the file is important or it is probably bad
-		vt.value = new fl::OutputVariable("Value");
-		vt.value->setMinimum(0);
-		vt.value->setMaximum(5);
-
-		std::vector<fl::InputVariable *> helper = {vt.strengthRatio, vt.heroStrength, vt.turnDistance, vt.missionImportance, vt.estimatedReward};
+		strengthRatio = new fl::InputVariable("strengthRatio"); //hero must be strong enough to defeat guards
+		heroStrength = new fl::InputVariable("heroStrength"); //we want to use weakest possible hero
+		turnDistance = new fl::InputVariable("turnDistance"); //we want to use hero who is near
+		missionImportance = new fl::InputVariable("lockedMissionImportance"); //we may want to preempt hero with low-priority mission
+		estimatedReward = new fl::InputVariable("estimatedReward"); //indicate AI that content of the file is important or it is probably bad
+		value = new fl::OutputVariable("Value");
+		value->setMinimum(0);
+		value->setMaximum(5);
+
+		std::vector<fl::InputVariable *> helper = { strengthRatio, heroStrength, turnDistance, missionImportance, estimatedReward };
 		for(auto val : helper)
 		{
-			vt.engine.addInputVariable(val);
+			engine.addInputVariable(val);
 		}
-		vt.engine.addOutputVariable(vt.value);
+		engine.addOutputVariable(value);
 
-		vt.strengthRatio->addTerm(new fl::Ramp("LOW", SAFE_ATTACK_CONSTANT, 0));
-		vt.strengthRatio->addTerm(new fl::Ramp("HIGH", SAFE_ATTACK_CONSTANT, SAFE_ATTACK_CONSTANT * 3));
-		vt.strengthRatio->setRange(0, SAFE_ATTACK_CONSTANT * 3);
+		strengthRatio->addTerm(new fl::Ramp("LOW", SAFE_ATTACK_CONSTANT, 0));
+		strengthRatio->addTerm(new fl::Ramp("HIGH", SAFE_ATTACK_CONSTANT, SAFE_ATTACK_CONSTANT * 3));
+		strengthRatio->setRange(0, SAFE_ATTACK_CONSTANT * 3);
 
 		//strength compared to our main hero
-		vt.heroStrength->addTerm(new fl::Ramp("LOW", 0.2, 0));
-		vt.heroStrength->addTerm(new fl::Triangle("MEDIUM", 0.2, 0.8));
-		vt.heroStrength->addTerm(new fl::Ramp("HIGH", 0.5, 1));
-		vt.heroStrength->setRange(0.0, 1.0);
+		heroStrength->addTerm(new fl::Ramp("LOW", 0.2, 0));
+		heroStrength->addTerm(new fl::Triangle("MEDIUM", 0.2, 0.8));

```

Comment:
If input is exactly 0.2, no term will be actually active

Reply:
Terms must cover entire range of values, and at least one rule must trigger for every combination.

Reply:
Terms of input variables must overlap, it this case at 0.2 no rule will be triggered (membership function is zero for all of them)